### PR TITLE
Derive DIT region from UK postcode when adding a company to Data Hub

### PR DESCRIPTION
**Pull request by paulgain**
## Description of change
When a user adds a D&B company to Data Hub the last thing that is required is to select the DIT region. This piece of work preselects the DIT region by making an API call (Hawk authentication) to the [Data Store Service](https://github.com/uktrade/data-store-service) (via Node.js) passing the company postcode to retrieve the DIT region.

| Required environment variables |
| ------------- |
| DATA_STORE_SERVICE_ACCESS_KEY_ID |
| DATA_STORE_SERVICE_SECRET_ACCESS_KEY |
| DATA_STORE_SERVICE_POSTCODE_TO_REGION_URL |

1. /companies
2. Select `Add company`
3. Select `United Kingdom`
4. Search for a company by name
5. Select a company from the results list
6. `Continue`
7. The DIT region will be selected based on the company postcode

dit-region-preselected

If the Data Store Service is given a postcode that isn't recognised the response will be empty (no error), in this scenario the UI will default to `-- Select DIT region --`.

unkown-postcode

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"

- [ ] Has the branch been rebased to master?
- [ ] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [ ] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: src/client/index.jsx [hunk 1]
```diff
@@ -77,6 +77,10 @@ import * as addInteractionFormState from '../apps/interactions/apps/add-interact
 import * as addInteractionFormTasks from '../apps/interactions/apps/add-interaction/client/tasks'
 import addInteractionFormReducer from '../apps/interactions/apps/add-interaction/client/reducer'
 
+import * as addCompanyState from '../apps/companies/apps/add-company/client/state'
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/client/index.jsx:80]: That pattern is verbose, over time the file will become cumbersome, we can shrink it down to a single line.

File: dev-stack.sample.env [hunk 2]
```diff
@@ -11,6 +11,9 @@ DATA_HUB_BACKEND_ACCESS_KEY_ID=frontend-key-id
 DATA_HUB_BACKEND_SECRET_ACCESS_KEY=frontend-key
 DATA_HUB_FRONTEND_ACCESS_KEY_ID=frontend-key-id
 DATA_HUB_FRONTEND_SECRET_ACCESS_KEY=frontend-key
+DATA_STORE_SERVICE_ACCESS_KEY_ID=data-store-service-id
+DATA_STORE_SERVICE_SECRET_ACCESS_KEY=data-store-service-key
+DATA_STORE_SERVICE_POSTCODE_TO_REGION_URL=https://lookInVault.com/
```
[the remark this answers is not in the document]
Comment (paulgain) [reply dev-stack.sample.env:16]: 👍

File: mock-stack.sample.env [hunk 3]
```diff
@@ -2,6 +2,9 @@ API_ROOT=http://mock-backend:8001
 ARCHIVED_DOCUMENTS_BASE_URL=https://lookInVault.com
 DATA_HUB_BACKEND_ACCESS_KEY_ID=frontend-key-id
 DATA_HUB_BACKEND_SECRET_ACCESS_KEY=frontend-key
+DATA_STORE_SERVICE_ACCESS_KEY_ID=data-store-service-id
+DATA_STORE_SERVICE_SECRET_ACCESS_KEY=data-store-service-key
+DATA_STORE_SERVICE_POSTCODE_TO_REGION_URL=https://lookInVault.com
```
[the remark this answers is not in the document]
Comment (paulgain) [reply mock-stack.sample.env:7]: 👍

File: visual-stack.sample.env [hunk 4]
```diff
@@ -6,6 +6,9 @@ TZ=Europe/London
 NODE_ENV=development
 DATA_HUB_BACKEND_ACCESS_KEY_ID=lookinvault
 DATA_HUB_BACKEND_SECRET_ACCESS_KEY=lookinvault
+DATA_STORE_SERVICE_ACCESS_KEY_ID=lookinvault
+DATA_STORE_SERVICE_SECRET_ACCESS_KEY=lookinvault
+DATA_STORE_SERVICE_POSTCODE_TO_REGION_URL=lookinvault
```
[the remark this answers is not in the document]
Comment (paulgain) [reply visual-stack.sample.env:11]: 👍

File: test/functional/cypress/specs/companies/add-company-spec.js [hunk 5]
```diff
@@ -587,7 +587,7 @@ describe('Add company form', () => {
 
     it('should add a company after defining both region and sector', () => {
       cy.get(selectors.companyAdd.regionSelect)
-        .select('London')
+        .select('South East')
```
[the remark this answers is not in the document]
Comment (paulgain) [reply test/functional/cypress/specs/companies/add-company-spec.js:595]: 👍

File: src/apps/api/controllers/postcode-to-region-lookup.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    if (results.length === 0) {
+      // Unknown postcode.
+      res.json({})
+      return
+    }
+
+    const regionName = results[0].regionName
+    const regions = await getOptions(token, 'uk-region')
+    const region = regions.filter((region) => region.label === regionName)
+    res.json(region[0])
```
Comment (rafenden) [inline src/apps/api/controllers/postcode-to-region-lookup.js:19]: You could use `Array.find`

```suggestion
    return regions.find((region) => region.label === regionName)
```

Comment (paulgain) [reply src/apps/api/controllers/postcode-to-region-lookup.js:19]: 👍

**Pull request merged**